#Student: Ana Flávia Martins dos Santos

##a - use the breast cancer classification dataset

In [ ]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import KFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, recall_score, precision_score
from sklearn.linear_model import LogisticRegression
import numpy as np

x, y = load_breast_cancer(return_X_y=True)

##b - use 5-fold cross validation

In [ ]:
kfold = KFold(n_splits=5)

##c - train and evaluate 3 different learning algorithms

##d - compare metrics for the 3 methods

In [ ]:
models = {
    "Random Forest": RandomForestClassifier(),
    "Logistic Regression": LogisticRegression(max_iter=10000),
    "KNN": KNeighborsClassifier()
}

for name, model in models.items():
    accuracy = []
    recall = []
    precision = []

    for train_index, test_index in kfold.split(x):
        X_train, X_test = x[train_index], x[test_index]
        y_train, y_test = y[train_index], y[test_index]

        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)

        accuracy.append(accuracy_score(y_test, y_pred))
        recall.append(recall_score(y_test, y_pred))
        precision.append(precision_score(y_test, y_pred))

    print(f"{name}:")
    print(f"Accuracy: {np.mean(accuracy):.4f}")
    print(f"Recall: {np.mean(recall):.4f}")
    print(f"Precision: {np.mean(precision):.4f}\n")

Random Forest:
Accuracy: 0.9561
Recall: 0.9705
Precision: 0.9492

Logistic Regression:
Accuracy: 0.9508
Recall: 0.9753
Precision: 0.9423

KNN:
Accuracy: 0.9262
Recall: 0.9652
Precision: 0.9128



##e - report experiment

Analyzing the following 3 accuracy, we can see that Random Forest has the highest accuracy, which indicates that it probably gets more predictions right (positive and negative)

Random Forest: 0.956

Logistic Regression: 0.950

KNN: 0.936



Analyzing the following 3 recall, we can see that Logistic Regression has the highest recall, which indicates that it is better predicting positive cases.

Random Forest: 0.970

Logistic Regression: 0.975

KNN: 0.965

Analyzing the following 3 precision, we can see that Random Forest has the highest precision, which indicates that it is more reliable predicting positive cases.

Random Forest: 0.949

Logistic Regression: 0.942

KNN: 0.912